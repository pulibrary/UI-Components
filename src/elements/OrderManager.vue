<template>
<div class="OrderManager">
  {{label}}
</div>
</template>

<script>
/**
 * OrderManager comes with its own Vuex store that it needs to manage state.
 * The easiest way to load the OrderManager is to simply pass a resource in as a prop.
 * You can try this below by adding `:resource="{id: 'foo', label: 'bar'}"` as an attribute/prop
 * to the `<ordermanager>` component.
 *
 * You can also use an API that will update the store. To do this, you
 * could wrap the OrderManager in a component that handles any syncing between the backend
 * and the Vuex store. Example: [todo]
 */
export default {
  name: "OrderManager",
  status: "review",
  release: "1.0.0",
  type: "Element",
  computed: {
    label() {
      return this.$store.getters.label
    },
  },
  props: {
    /**
     * The html element name used for the nav bar.
     */
    resource: {
      type: Object,
      default: null,
    },
  },
  mounted() {
    if (this.resource) {
      this.$store.commit("SET_RESOURCE", this.resource)
    }
  },
}
</script>
<docs>
Don't forget that you can debug it with [vue-devtools](https://github.com/vuejs/vue-devtools)

```js
    <order-manager></order-manager>
```
</docs>
