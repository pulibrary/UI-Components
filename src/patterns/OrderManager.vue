<template>
<div class="orderManager">
  <transition name="fade">
    <div v-if="loading" class="overlay">
      <loader size="medium"></loader>
    </div>
  </transition>
  <heading level="h2">{{resource.label}} <small v-if="isMultiVolume">Multi-volume</small></heading>
  <wrapper class="galleryPanel" type="div">
    <gallery :galleryItems="galleryItems"></gallery>
  </wrapper>
  <wrapper class="sidePanel" type="div" :fullWidth="false">
    <!-- Resource Form-->
    <resource-form v-if="selectedTotal === 0"></resource-form>
    <!-- Multiple Selected Form-->
    <filesets-form v-if="selectedTotal > 1"></filesets-form>
    <!-- Single Selected Form-->
    <fileset-form v-if="selectedTotal === 1"></fileset-form>
  </wrapper>
</div>
</template>

<script>
import { mapState, mapGetters } from "vuex"
/**
 * OrderManager is a tool for reordering thumbnails that represent members of a complex object (a book, CD, multi-volume work, etc.).
 * Complex patterns like OrderManager come with their own Vuex store that it needs to manage state.
 * The easiest way to use the OrderManager is to simply pass a resource in as a prop.
 * You can see how this is done in the live code example at the end of this section.
 *
 * However you will still need to load the corresponding
 * Vuex module, *resourceModule*. Please see [the state management documentation](/#!/State%20Management) for how to manage state in complex patterns.
 */
export default {
  name: "OrderManager",
  status: "review",
  release: "1.0.0",
  type: "Element",
  computed: {
    galleryItems() {
      return this.resource.members.map(member => ({
        id: member.id,
        title: member.label,
        caption: member.__typename + " : " + member.id,
        mediaUrl: "https://picsum.photos/600/300/?random",
        viewingHint: member.viewingHint,
      }))
    },
    selectedTotal() {
      return this.gallery.selected.length
    },
    isMultiVolume() {
      return this.$store.getters.isMultiVolume
    },
    ...mapState({
      resource: state => state.ordermanager.resource,
      gallery: state => state.gallery,
    }),
    loading: function() {
      return this.resource.loadState !== "LOADED" ? true : false
    },
  },
  props: {
    /**
     * The resource object in json format. Example: [todo]
     */
    resourceObject: {
      type: Object,
      default: null,
    },
  },
  mounted() {
    if (this.resourceObject) {
      // if props are passed in set the resource on mount
      this.$store.commit("SET_RESOURCE", this.resourceObject)
    } else {
      this.$store.commit("CHANGE_RESOURCE_LOAD_STATE", "LOADING")
    }
  },
}
</script>
<style lang="scss" scoped>
.title {
  font-weight: bold;
}
.orderManager {
  position: relative;
  height: 80vh;
}
.sidePanel {
  position: absolute;
  top: 20px;
  right: 10px;
  height: 95%;
  width: 28.5%;
  border: 1px solid #ddd;
  border-radius: 4px;

  padding: 0 30px 0 30px;
  // height: 100%;
  overflow-y: scroll;
}
.galleryPanel {
  position: absolute;
  top: 20px;
  left: 0;
  height: 95%;
  width: 70%;
  border-radius: 4px;
  border: 1px solid #ddd;
}
</style>
<docs>
<br/>Don't forget that you can debug it with [vue-devtools](https://github.com/vuejs/vue-devtools)

```js
    <order-manager :resource-object='{
      "id": "aea40813-e0ed-4307-aae9-aec53b26bdda",
      "label": "Resource with 3 files",
      "viewingHint": "individuals",
      "__typename": "ScannedResource",
      "members": [
        {
          "id": "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
          "label": "a",
          "viewingHint": null,
          "__typename": "FileSet"
        },
        {
          "id": "8f0a0908-317f-414e-a78a-c38a4a3b28e3",
          "label": "b",
          "viewingHint": null,
          "__typename": "FileSet"
        },
        {
          "id": "ea01019e-f644-4416-b99c-1b44bf49d060",
          "label": "c",
          "viewingHint": null,
          "__typename": "FileSet"
        }
      ]
    }'></order-manager>
```
</docs>
