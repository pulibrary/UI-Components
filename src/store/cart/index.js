export const cartState = {
  items: [],
  isVisible: false,
}

export const cartActions = {
  addItemToCart(context, newItem) {
    const duplicate = context.state.items.find(item => item.callnumber === newItem.callnumber)
    console.log(context)
    if (typeof duplicate === "undefined") {
      context.commit("PUSH_ITEM_TO_CART", newItem)
    }
  },
}

export const cartMutations = {
  TOGGLE_VISIBILITY(state) {
    state.isVisible = !state.isVisible
  },
  SET_CART(state, items) {
    state.items = items
  },
  PUSH_ITEM_TO_CART(state, item) {
    state.items.push(item)
  },
}
