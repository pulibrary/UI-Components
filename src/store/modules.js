export const counterModule = {
  state: { count: 10 },

  mutations: {
    increment(state) {
      // `state` is the local module state
      state.count++
    },
  },

  getters: {
    count: state => {
      return state.count
    },
  },
}

export const galleryModule = {
  // namespaced: true,

  state: {
    items: [],
    selected: [],
    cut: [],
    changeList: [],
    ogItems: [],
  },

  mutations: {
    CUT(state, itemArray) {
      state.cut = [...itemArray]
    },
    PASTE(state, itemArray) {
      state.items = [...itemArray]
    },
    SELECT(state, itemArray) {
      state.selected = [...itemArray]
    },
    SET_GALLERY(state, items) {
      state.items = items
      state.ogItems = items
    },
    SORT_ITEMS(state, value) {
      state.items = [...value]
    },
    UPDATE_CHANGES(state, changeList) {
      state.changeList = [...changeList]
    },
    UPDATE_ITEMS(state, items) {
      state.items = [...items]
    },
  },
}

export const resourceModule = {
  state: {
    resource: {
      id: "",
      resourceClassName: "",
      bibId: "",
      label: "Resource not available.",
      thumbnail: "",
      startCanvas: "",
      isMultiVolume: false,
      viewingHint: null,
      viewingDirection: null,
      members: [],
      loadState: "NOT_LOADED",
      ogState: {},
    },
  },

  mutations: {
    CHANGE_RESOURCE_LOAD_STATE(state, loadState) {
      state.resource.loadState = loadState
    },
    SET_RESOURCE(state, resource) {
      state.resource.id = resource.id
      state.resource.resourceClassName = resource.__typename
      state.resource.label = resource.label
      state.resource.startCanvas = resource.startPage
      state.resource.viewingHint = resource.viewingHint
      state.resource.viewingDirection = resource.viewingDirection
      state.resource.thumbnail = resource.thumbnail.id
      state.resource.members = resource.members
      const items = resource.members.map(member => ({
        id: member.id,
        viewingHint: member.viewingHint != null ? member.viewingHint : "single",
        caption: member.label, // member.__typename + " : " + member.id,
        service:
          typeof member.thumbnail.iiifServiceUrl != "undefined"
            ? member.thumbnail.iiifServiceUrl
            : "https://picsum.photos/600/300/?random",
        mediaUrl:
          typeof member.thumbnail.iiifServiceUrl != "undefined"
            ? member.thumbnail.iiifServiceUrl + "/full/300,/0/default.jpg"
            : "https://picsum.photos/600/300/?random",
      }))
      state.gallery.items = items
      state.gallery.ogItems = items
      state.resource.loadState = "LOADED"
      state.resource.ogState = {
        startCanvas: resource.startPage,
        thumbnail: resource.thumbnail,
        viewingHint: resource.viewingHint,
        viewingDirection: resource.viewingDirection,
      }
    },
    UPDATE_STARTCANVAS(state, startCanvas) {
      state.resource.startCanvas = startCanvas
    },
    UPDATE_THUMBNAIL(state, thumbnail) {
      state.resource.thumbnail = thumbnail
    },
    UPDATE_VIEWDIR(state, viewDir) {
      state.resource.viewingDirection = viewDir
    },
    UPDATE_VIEWHINT(state, viewHint) {
      state.resource.viewingHint = viewHint
    },
  },

  // actions: any async actions are should be handled by the host app

  getters: {
    getMemberCount: state => {
      return state.resource.members.length
    },
    isMultiVolume: state => {
      const volumes = state.resource.members.filter(member => member.__typename === "ScannedResource")
      return volumes.length > 0 ? true : false
    },
    orderChanged: state => {
      let ogOrder = JSON.stringify(state.gallery.ogItems.map(item => item.id))
      let imgOrder = JSON.stringify(state.gallery.items.map(item => item.id))
      return ogOrder !== imgOrder
    },
    stateChanged: (state, getters) => {
      var propsChanged = []
      propsChanged.push(state.resource.ogState.thumbnail !== state.thumbnail)
      propsChanged.push(state.resource.ogState.startPage !== state.startPage)
      propsChanged.push(state.resource.ogState.viewingHint !== state.viewingHint)
      propsChanged.push(state.resource.ogState.viewingDirection !== state.viewingDirection)
      propsChanged.push(state.gallery.changeList.length > 0)
      propsChanged.push(getters.orderChanged)
      if (propsChanged.indexOf(true) > -1) {
        return true
      } else {
        return false
      }
    },
  },
  modules: {
    gallery: galleryModule,
  },
}

let modules
export default (modules = { counterModule, galleryModule, resourceModule })
