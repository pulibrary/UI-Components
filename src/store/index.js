import Vue from "vue"
import Vuex from "vuex"
import { counterModule, galleryModule } from "./modules"

Vue.use(Vuex)

export default new Vuex.Store({
  modules: {
    counter: counterModule,
    gallery: galleryModule,
  },
})
