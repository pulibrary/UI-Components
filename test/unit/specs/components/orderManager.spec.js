import Vuex from "vuex"
import { createLocalVue, mount, shallowMount } from "@vue/test-utils"
import OrderManager from "@/patterns/OrderManager.vue"
import { resourceModule, galleryModule } from "@/store/modules"

// create an extended `Vue` constructor
const localVue = createLocalVue()
localVue.use(Vuex)

let wrapper
let store
let galleryState
let state
let mutations
let items = [
  { id: "a", title: "A", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
]
let resourceObject = {
  id: "aea40813-e0ed-4307-aae9-aec53b26bdda",
  label: "Resource with 3 files",
  viewingHint: "individuals",
  viewingDirection: "LEFTTORIGHT",
  startPage: "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
  thumbnail: {
    id: "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
    thumbnailUrl:
      "http://localhost:3000/image-service/8ffd7a03-ec0e-46c1-a347-e4b19cb7839f/full/!200,150/0/default.jpg",
    iiifServiceUrl: "http://localhost:3000/image-service/8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
  },
  __typename: "ScannedResource",
  members: [
    {
      id: "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
      label: "a",
      viewingHint: null,
      thumbnail: {
        iiifServiceUrl:
          "https://libimages1.princeton.edu/loris/figgy_prod/f7%2F67%2Ffe%2Ff767fe4247524c5f96e16eba2ff93301%2Fintermediate_file.jp2",
      },
      __typename: "FileSet",
    },
    {
      id: "8f0a0908-317f-414e-a78a-c38a4a3b28e3",
      label: "b",
      viewingHint: null,
      thumbnail: {
        iiifServiceUrl:
          "https://libimages1.princeton.edu/loris/figgy_prod/a5%2F15%2F62%2Fa515627580594c978ce5352653c9442a%2Fintermediate_file.jp2",
      },
      __typename: "FileSet",
    },
    {
      id: "ea01019e-f644-4416-b99c-1b44bf49d060",
      label: "c",
      viewingHint: null,
      thumbnail: {
        iiifServiceUrl:
          "https://libimages1.princeton.edu/loris/figgy_prod/d9%2Fb5%2F8c%2Fd9b58c8f3e554706bec4d977b12cd4e4%2Fintermediate_file.jp2",
      },
      __typename: "FileSet",
    },
  ],
}
//
describe("OrderManager.vue", () => {
  //   beforeEach(() => {
  //     galleryState = {
  //       items: items,
  //       selected: [items[0]],
  //       cut: [],
  //       changeList: ["2"],
  //       ogItems: items,
  //     }
  //
  //     state = {
  //       id: "",
  //       resourceClassName: "",
  //       bibId: "",
  //       label: "Resource not available.",
  //       thumbnail: "",
  //       startCanvas: "",
  //       isMultiVolume: false,
  //       viewingHint: null,
  //       viewingDirection: null,
  //       members: [],
  //       loadState: "NOT_LOADED",
  //       saveState: "NOT_SAVED",
  //       ogState: {},
  //     }
  //
  //     mutations = {
  //       SET_RESOURCE: jest.fn(),
  //     }
  //
  //     store = new Vuex.Store({
  //       modules: {
  //         ordermanager: {
  //           state
  //         },
  //         // gallery: {
  //         //   galleryState
  //         // },
  //       },
  //     })
  //
  //     // const Card = createStub("Card")
  //     wrapper = mount(OrderManager, {
  //       localVue,
  //       store,
  //       propsData: {
  //         resourceObject: resourceObject,
  //       },
  //       // stubs: ["card", "heading", "text-style", "media-image"],
  //     })
  //   })

  it("slug", () => {
    expect(true).toBe(true)
  })

  // it("has the expected html structure", () => {
  //   expect(wrapper.element).toMatchSnapshot()
  // })
})
