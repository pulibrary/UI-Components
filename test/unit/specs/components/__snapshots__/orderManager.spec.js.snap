// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderManager.vue has the expected html structure 1`] = `
<div
  class="orderManager"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <wrapper-stub
    class="galleryPanel"
    fullwidth="true"
    type="div"
  >
    <toolbar-stub />
     
    <gallery-stub
      cardpixelwidth="300"
      class="galleryWrapper"
      galleryitems="[object Object],[object Object],[object Object]"
    />
  </wrapper-stub>
   
  <wrapper-stub
    class="sidePanel"
    type="div"
  >
    <!---->
     
    <!---->
     
    <fileset-form-stub />
     
    <controls-stub
      viewerid="viewer"
    />
  </wrapper-stub>
</div>
`;
